### ISSUE TITLE
Access Error 403 - Access Denied from Curl in a C program, but not from shell

### ISSUE THREAD

**Issue by gpu-1** [Voyz/ibeam#183]
Before anything, I want to thank you for a great program. It will make my life much simpler once I have it running. 

I am having the following problem. running Docker image with custom SSL certs.

To run:
sudo sudo docker run --env-file ~/IBK/gw/env.list -p 5000:5000 --name ibeam -v /home/IBK/gw/root:/srv/inputs/ voyz/ibeam
2024-05-05 17:34:42,793|I| ############ Starting IBeam version 0.5.1 ############
2024-05-05 17:34:42,794|I| TLS certificates found and will be used for verification
2024-05-05 17:34:42,794|I| Custom conf.yaml found and will be used by the Gateway
2024-05-05 17:34:42,796|I| Secrets source: env
2024-05-05 17:34:42,797|I| Health server started at port=5001

On loading:

2024-05-05 17:35:22,048|I| Loading auth webpage at https://localhost:5000/sso/Login?forwardTo=22&RL=1&ip2loc=on
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by io.netty.util.internal.ReflectionUtil (file:/srv/clientportal.gw/build/lib/runtime/netty-common-4.1.15.Final.jar) to constructor java.nio.DirectByteBuffer(long,int)
WARNING: Please consider reporting this to the maintainers of io.netty.util.internal.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
 -> mount demo on /demo
Java Version: 11.0.20
****************************************************
version: a27ed42161ad96c53e715ca5c5e3e3fa4cff5262 Mon, 24 Apr 2023 15:41:53 -0400
****************************************************

then it goes in authentication

Logging in succeeded
Repeating status check attempts another 14 times
AUTHENTICATED Status(running=True, session=True, connected=True, authenticated=True, competing=False, collision=False, session_id='411dc6xxxxxxx85acab', server_name=None, server_version=None, expires=593498)
Gateway running and authenticated, session id: 411dc680e89xxxxxxx185acab, server name: None

from command line:
curl -X GET "https://localhost:5000/v1/api/portfolio/accounts" --cacert cacert.pem

response is as expected:
[{"id":"U13sssdddd",....................."desc":"U138dfdfdf"}]

I wrote a c program that uses the same curl endpoint:

      curl = curl_easy_init();
      if (curl) {
      curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "GET");
      curl_easy_setopt(curl, CURLOPT_URL, "https://127.0.0.1:5000/v1/api/portfolio/accounts");
      /// to get response from server
      curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
      curl_easy_setopt(curl, CURLOPT_CAINFO,"/home/IBK/gw/root/cacert.pem");
      curl_easy_setopt(curl, CURLOPT_WRITEDATA, response_buffer);
      curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L); 

       curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
       res = curl_easy_perform(curl);
       curl_easy_cleanup(curl);
       }

when the program is executed, the docker gw  connects, authenticates the SSL. then rejects the request with 403:

*   Trying 127.0.0.1:5000...
* Connected to 127.0.0.1 (127.0.0.1) port 5000 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /home/IBK/gw/root/cacert.pem
*  CApath: /etc/ssl/certs
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=US; ST=FL; L=Miami; O=Andres; CN=Andres
*  start date: May  4 15:44:40 2024 GMT
*  expire date: May  4 15:44:40 2026 GMT
*  subjectAltName: host "127.0.0.1" matched cert's IP address!
*  issuer: C=US; ST=FL; L=Miami; O=Andres; CN=Andres
*  SSL certificate verify ok.
> GET /v1/api/iserver/accounts HTTP/1.1
Host: 127.0.0.1:5000
Accept: */*

* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< Referrer-Policy: Origin-when-cross-origin
< Content-Type: text/html; charset=iso-8859-1
< Expires: Sun, 05 May 2024 17:35:56 GMT
< Cache-Control: max-age=0, no-cache, no-store
< Pragma: no-cache
< Date: Sun, 05 May 2024 17:35:56 GMT
< Connection: keep-alive
< Set-Cookie: x-sess-uuid=0.1b213017.17156.b8d0225; secure; HttpOnly
< Server-Timing: cdn-cache; desc=MISS
< Server-Timing: edge; dur=4
< Server-Timing: origin; dur=33
< Strict-Transport-Security: max-age=600 ; includeSubDomains
< Server-Timing: ak_p; desc="1714930556511_389030171_193790501_3785_7176_20_0_-";dur=1
< Transfer-Encoding: chunked
< 
* Connection #0 to host 127.0.0.1 left intact
Resp Orig v6: Error 403 - Access Denied
Resp1: Error 403 - Access Denied 

my conf.yaml:

    ip2loc: "US"
    proxyRemoteSsl: true
    proxyRemoteHost: "https://api.ibkr.com"
    listenPort: 5000
    listenSsl: true
    ccp: false
    svcEnvironment: "v1"
    sslCert: "cacert.jks"
    sslPwd: "mypass"
    authDelay: 3000
    portalBaseURL: ""
    serverOptions:
        blockedThreadCheckInterval: 1000000
        eventLoopPoolSize: 20
        workerPoolSize: 20
        maxWorkerExecuteTime: 100
        internalBlockingPoolSize: 20
    cors:
        origin.allowed: "*"
        allowCredentials: false
    webApps:
        - name: "demo"
          index: "index.html"
    ips:
      allow:
        - 192.*
        - 131.216.*
        - 127.0.0.1
        - 0.0.0.0
        - 172.17.0.0
        - 172.17.0.1
        - 172.17.0.2
        - 172.17.0.3
        - 192.xxx.xxx.xxx  // my ip
      deny:
        - 212.90.324.10
       
Ibeam continues running and stays connected. 

curl is running the same version from the shell and from the c program:
 libcurl version: 7.81.0
system :   Ubuntu 22.04.4 LTS (GNU/Linux 5.15.0-105-generic x86_64)
Java Version: 11.0.20 on docker

BTY:
-same error results when executing as root.
-same error when trying other endpoints.

Thanks for your help

**Comment by Voyz** [comment]
Hey @gpu-1 thanks for checking out IBeam and for your kind words 😊

As for your issue, I cannot see anything that would directly point to the cause. I've never used C/curl to communicate with the Gateway in the way you do. That being said, I do use curl successfully on its own, and I can see that it is working for you just fine.

Additionally, I'm noticing that you both provide the SSL certificate and disable SSL certificate verification (I'm deducing this from your comments). 

```
  curl_easy_setopt(curl, CURLOPT_CAINFO,"/home/IBK/gw/root/cacert.pem");
  ...
  curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1L); // Disable SSL certificate verification
```

Possibly this could cause some issues?

In either case, if you disable the SSL verification, why the log report:

```
SSL certificate verify ok.
```

Sounds like possibly something isn't quite correct in the setup. 

If you access the the IBeam container you can find the Gateway logs in the clientportal.gw/logs directory. See if there are any clues there.

**Comment by gpu-1** [comment]
I am sorry. that commented (// Disable SSL cert verif) was from a previous test where I blocked SSL verification. Its is actually verifying SSL. I edited my original post  to clarify

I followed your advise and logged into the GW log.
this is what I see when i test with the C curl:

04:22:44.737 INFO  vert.x-eventloop-thread-0 BaseServiceProxy     : -> GET /v1/api/iserver/accounts,403|121ms

    response is 403.

this is the log of the shell curl:

04:29:12.797 INFO  vert.x-eventloop-thread-0 BaseServiceProxy     : -> GET /v1/api/portfolio/accounts,200|94ms

   response is 200.
